src/group_theory/perm/basic.lean

```diff
@@ -234,29 +236,66 @@ lemma extend_domain_hom_injective : function.injective (extend_domain_hom f) :=
 
 end extend_domain
 
+section subtype
+variables {p : α → Prop}
+
 /-- If the permutation `f` fixes the subtype `{x // p x}`, then this returns the permutation
   on `{x // p x}` induced by `f`. -/
-def subtype_perm (f : perm α) {p : α → Prop} (h : ∀ x, p x ↔ p (f x)) : perm {x // p x} :=
+def subtype_perm (f : perm α) (h : ∀ x, p x ↔ p (f x)) : perm {x // p x} :=
 ⟨λ x, ⟨f x, (h _).1 x.2⟩, λ x, ⟨f⁻¹ x, (h (f⁻¹ x)).2 $ by simpa using x.2⟩,
   λ _, by simp only [perm.inv_apply_self, subtype.coe_eta, subtype.coe_mk],
   λ _, by simp only [perm.apply_inv_self, subtype.coe_eta, subtype.coe_mk]⟩
 
-@[simp] lemma subtype_perm_apply (f : perm α) {p : α → Prop} (h : ∀ x, p x ↔ p (f x))
+@[simp] lemma subtype_perm_apply (f : perm α) (h : ∀ x, p x ↔ p (f x))
   (x : {x // p x}) : subtype_perm f h x = ⟨f x, (h _).1 x.2⟩ := rfl
 
-@[simp] lemma subtype_perm_one (p : α → Prop) (h : ∀ x, p x ↔ p ((1 : perm α) x)) :
-  @subtype_perm α 1 p h = 1 :=
+@[simp] lemma subtype_perm_one (p : α → Prop) (h := λ _, iff.rfl) : @subtype_perm α p 1 h = 1 :=
 equiv.ext $ λ ⟨_, _⟩, rfl
 
+@[simp] lemma subtype_perm_mul (f g : perm α) (hf hg) :
+  (f.subtype_perm hf * g.subtype_perm hg : perm {x // p x}) =
+    (f * g).subtype_perm (λ x, (hg _).trans $ hf _) := rfl
+
+private lemma inv_aux {f : perm α} (hf : ∀ x, p x ↔ p (f x)) (x : α) : p x ↔ p (f⁻¹ x) :=
+by rw [hf (f⁻¹ x), f.apply_inv_self]
+
+private lemma pow_aux {f : perm α} (hf : ∀ x, p x ↔ p (f x)) : ∀ {n : ℕ} x, p x ↔ p ((f ^ n) x)
+| 0 x := iff.rfl
+| (n + 1) x := (pow_aux _).trans (hf _)
+
+private lemma zpow_aux {f : perm α} (hf : ∀ x, p x ↔ p (f x)) : ∀ {n : ℤ} x, p x ↔ p ((f ^ n) x)
+| (int.of_nat n) := pow_aux hf
+| (int.neg_succ_of_nat n) := by { rw zpow_neg_succ_of_nat, exact inv_aux (pow_aux hf) }
```

Comment:
You could save yourself from having to prove these by creating a `private` `subgroup (perm α)` with `carrier := {f | ∀ x, p x ↔ p (f x))}`

Reply:
Unfortunately, `subgroup` imports this file.

Reply:
You could move these results to `equiv/perm/subgroup`? Then you could even make this a true monoid_hom.

Reply:
I see little point in doing so. The auxiliary lemmas would get golfed overall by at most 10 lines and `subtype_perm` *cannot* be written as a `perm α →* perm {x // p x}`. You would need to restrict to the subgroup of perms `f` respecting `∀ x, p x ↔ p (f x))`, which is `perm {x // p x}` anyway, and at which point you've lost all use of the lemmas I'm introducing here.

Reply:
This is the meaning of my second sentence. The map you're talking about is not `subtype_perm` anymore, so I have no use for it.

Reply:
Yes, absolutely. But I'm interested in rewriting, not showing the isomorphism between `perm {x // p x}` and a subgroup of `perm α`.